crates/sui-graphql-rpc/src/types/object.rs

```diff
@@ -340,3 +386,142 @@ fn addr(bytes: impl AsRef<[u8]>) -> Result<SuiAddress, Error> {
+    fields: VecMap<String, String>,
+    move_struct: &MoveStruct,
+) -> Result<Vec<DisplayEntry>, Error> {
+    let mut rendered_fields = Vec::<DisplayEntry>::new();
+
+    for entry in fields.contents.iter() {
+        let rendered_value = parse_template(&entry.value, move_struct)?;
+        rendered_fields.push(DisplayEntry {
+            key: entry.key.clone(),
+            value: rendered_value,
+        });
+    }
+
+    Ok(rendered_fields)
+}
+
+// handles the PART = '{' CHAIN '}'
+fn parse_template(template: &str, move_struct: &MoveStruct) -> Result<String, Error> {
+    let mut output = template.to_string();
+    let mut var_name = String::new();
+    let mut in_braces = false;
+    let mut escaped = false;
+
+    for ch in template.chars() {
+        match ch {
+            '\\' => {
+                escaped = true;
+                continue;
+            }
+            '{' if !escaped => {
+                in_braces = true;
+                var_name.clear();
+            }
+            '}' if !escaped => {
+                in_braces = false;
+                let value = get_value_from_move_struct(move_struct, &var_name)?;
+                output = output.replace(&format!("{{{}}}", var_name), &value.to_string());
+            }
+            _ if !escaped => {
+                if in_braces {
+                    var_name.push(ch);
+                }
+            }
+            _ => {}
+        }
+        escaped = false;
+    }
+
+    Ok(output.replace('\\', ""))
+}
+
+pub fn get_value_from_move_struct(
+    move_struct: &MoveStruct,
+    var_name: &str,
+) -> Result<String, Error> {
+    // Supports CHAIN . IDENT today
+    let parts: Vec<&str> = var_name.split('.').collect();
+    if parts.is_empty() {
+        // todo: custom error
+        Err(Error::Internal(
+            "Display template value cannot be empty".to_string(),
+        ))?;
+    }
+    // todo: new limit on config
+    if parts.len() > 10 {
+        Err(Error::Internal(format!(
+            "Display template value nested depth cannot exist {}",
+            10
+        )))?;
+    }
+
+    // update this as we iterate through the parts
+    let start_value = &MoveValue::Struct(move_struct.clone());
+
+    let result = parts
+        .iter()
+        .try_fold(start_value, |current_value, part| match current_value {
+            MoveValue::Struct(s) => s
+                .fields
+                .iter()
+                .find_map(|(id, value)| {
+                    if id.to_string() == *part {
+                        Some(value)
+                    } else {
+                        None
+                    }
+                })
+                .ok_or_else(|| Error::Internal(format!("Field '{}' not found", part))),
+            _ => Err(Error::Internal("Unexpected MoveValue".to_string())),
+        })?;
+
+    // TODO: implement Display for MoveData and use that instead
+    let sui_move_value: SuiMoveValue = result.clone().into();
```

Comment:
Instead of implementing `Display` on `MoveData`, we can adapt the `Display` value field to be of type `Json`, and use the existing `MoveValue::json_impl` to convert into the appropriate scalar.

Reply:
I see, that's pretty neat

Reply:
Okay, this would be pretty neat if we didn't need to substitute values into the template. But since we do, I end up having to do a lot more conversions, from MoveValue -> graphql::Json -> String, interpolate this into the template, then back into graphql::Json. That said, I think this would be okay if we're okay with returning DisplayEntry with fields key, value -> value to be interpolated, template instead of the existing functionality on jsonrpc, which is that the rpc returns the value w/ the string interpolated.

If not, it looks like when we interpolate the move value into the string, it keeps the json string escape symbols. I'll handle this in a subsequent PR, and this one will focus more on setting up e2e tests that we can depend on later